### PR TITLE
Clifford decomposition doc

### PR DESCRIPTION
**Pull request by asciineuron**
# Description

### CODE REVIEW

**Review by rmlarose — COMMENTED** [review]
I don't understand the argument for why the last plot is interesting and would remove this. Can you increase the repetitions in the first two plots so that the histograms match better?

**Comment by asciineuron** [comment]
I added seeds so it's more reproducible and got the plots looking better!
The last plot is kind of the point of the decomposition function in the first place. If you have a list of pure-Clifford circuits approximating the original, you can run each exactly with a fast CliffordSim and then add the state vectors. This could be faster than exactly simulating the original circuit with some slower simulation method.

**Comment by asciineuron** [comment]
Did the same thing here and merged main into this branch. One additional change: in `simulator.py` here I changed into a loop over moments and then operators in the given moment. For random circuits (as opposed to handcrafted ones) there can be many operators in a moment, leading to unexpected behavior in ensuring terminal measurement.

**Comment by rmlarose** [comment]
> For random circuits (as opposed to handcrafted ones) there can be many operators in a moment, leading to unexpected behavior in ensuring terminal measurement.

So the check for terminal measurements was buggy? If so please add a unit test that used to fail and now passes with the change

**Comment by asciineuron** [comment]
@rmlarose Hey Ryan - this new test should highlight it. Basically, circuits can have many ops per moment. Before we checked if a measurement is terminal by indexing the moment number with the operation number - which only fully works if its 1-to-1.

I think we never saw this before because repeatedly measuring all the qubits, or using the same key name, overrides the prior measurement (since we're only considering the final terminal one). So it only changes results for partial measurements like in the example.

File: bgls/simulator_test.py [hunk 1]
```diff
@@ -352,3 +352,39 @@ def test_mps_results_match_state_vec():
...
+
+def test_dense_moments_intermediate_measurement_counting():
+    """For a circuit with several operations per moment, counting operations
+    rather than moments can erroneously classify an intermediate measurement
+    as terminal"""
+    q = cirq.LineQubit.range(3)
+    circuit1 = cirq.Circuit(
+        (
+            cirq.Moment(cirq.X.on(q[0]), cirq.X.on(q[1]), cirq.X.on(q[2])),
+            cirq.Moment(cirq.X.on(q[0]), cirq.X.on(q[1]), cirq.X.on(q[2])),
```
Comment (rmlarose) [inline bgls/simulator_test.py:365]: Consider `cirq.X.on_each`

File: bgls/simulator_test.py [hunk 2]
```diff
@@ -352,3 +352,39 @@ def test_mps_results_match_state_vec():
     result_mps = sim_mps.run(circuit, repetitions=100)
 
     assert result_mps == result_state_vector
+
+
+def test_dense_moments_intermediate_measurement_counting():
```
Comment (rmlarose) [inline bgls/simulator_test.py:357]: How about `test_intermediate_measurements_are_ignored`? I don't know what you mean by counting

**Review by rmlarose — COMMENTED** [review]
Thanks for the test - two comments on this then please fix merge conflicts and merge. I want to clean up the notebooks a little but will just do this in a separate PR after this is merged.

**Pull request merged**